pkg/controller/deployment/deployment_controller.go

```diff
@@ -816,30 +834,102 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxCleanupCount int) (int, error) {
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+	// Safely scale down all old rcs with unhealthy replicas. ReplicationController/ReplicaSet will sort the pods in the order
+	// such that not-ready < ready, unscheduled < scheduled, and pending < running. This ensures that unhealthy replicas will
+	// been deleted first and won't increase unavailability.
+	totalScaledDown := 0
+	for _, targetRC := range oldRCs {
+		if totalScaledDown >= maxCleanupCount {
+			break
+		}
+
+		if targetRC.Spec.Replicas == 0 {
+			// cannot scale down this RC.
+			continue
+		}
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, 0)
+		if err != nil {
+			return totalScaledDown, fmt.Errorf("could not find available pods: %v", err)
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		scaledDown := int(math.Min(float64(maxCleanupCount-totalScaledDown), float64(targetRC.Spec.Replicas-readyPodCount)))
+		newReplicasCount := targetRC.Spec.Replicas - scaledDown
+		_, err = dc.scaleRCAndRecordEvent(targetRC, newReplicasCount, deployment)
+		if err != nil {
+			return totalScaledDown, err
+		}
+		totalScaledDown += scaledDown
+	}
+	return totalScaledDown, nil
+}
+
+// scaleDownOldRCsForRollingUpdate scales down old rcs when deployment strategy is "RollingUpdate".
+// Need check maxUnavailable to ensure availability
+func (dc *DeploymentController) scaleDownOldRCsForRollingUpdate(allRCs []*api.ReplicationController, oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxScaleDownCount int) (int, error) {
+	maxUnavailable, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxUnavailable)
+	if err != nil {
+		return 0, fmt.Errorf("invalid value for MaxUnavailable: %v", err)
+	}
+	if isPercent {
+		maxUnavailable = util.GetValueFromPercent(maxUnavailable, deployment.Spec.Replicas)
+	}
+	// Check if we can scale down.
+	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	minReadySeconds := deployment.Spec.MinReadySeconds
 	// Find the number of ready pods.
 	readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, minReadySeconds)
 	if err != nil {
-		return false, fmt.Errorf("could not find available pods: %v", err)
+		return 0, fmt.Errorf("could not find available pods: %v", err)
 	}
-
 	if readyPodCount <= minAvailable {
 		// Cannot scale down.
-		return false, nil
+		return 0, nil
 	}
-	totalScaleDownCount := readyPodCount - minAvailable
+
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+
 	totalScaledDown := 0
+	totalScaleDownCount := int(math.Max(float64(readyPodCount-minAvailable), float64(maxScaleDownCount)))
```

Comment:
Sorry, it's incorrect. It should be Min, instead of Max.

I'd like the total scale down count(unhealthy replicas cleaned up by `cleanupUnhealthyReplicas` +  scaled down by `scaleDownOldRCsForRollingUpdate`) doesn't exceed `maxScaledDown` (`allPodsCount - minAvailable - newRCPodsUnavailable`)